README.md


[![Greenkeeper badge](https://badges.greenkeeper.io/anydown/mangadown.svg)](https://greenkeeper.io/)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.2.0 (2020-08-21)


### Features

* add chinese font ([130d47f](https://github.com/anydown/mangadown/commit/130d47f317f803ac7fa220fdf57e2078c28d70af))
* add freedrawing mode ([5d3b84f](https://github.com/anydown/mangadown/commit/5d3b84f38c95c5c18b106e3ae59666fad4b8f5c4))
* add local image loader ([c3e2c23](https://github.com/anydown/mangadown/commit/c3e2c23a9381ee84c6f866ba9169cff1160fa6d4))
* fill white inner bubbles ([42b9ff8](https://github.com/anydown/mangadown/commit/42b9ff84d5bcbadc028ae938bd238545cfda3119))


### Bug Fixes

* Add support devicePixelRatio > 1 ([3d939d5](https://github.com/anydown/mangadown/commit/3d939d598408107d86000f700145b2f187b9b964))
* change title ([74182b0](https://github.com/anydown/mangadown/commit/74182b00edcca7939d093a067a849efa1296bc74))
* discard selection when export image ([344a482](https://github.com/anydown/mangadown/commit/344a482549a9ddacc1214b3e2db7df8b4f3e99d9))
* fix image url ([fe056ed](https://github.com/anydown/mangadown/commit/fe056edd4966d5d37c5e0535d7324e474fcf4c11))
* remove multiple objects ([c41b7b5](https://github.com/anydown/mangadown/commit/c41b7b547f5b33a1da015ecdea13e902571d7980))
* save as PNG ([e62630e](https://github.com/anydown/mangadown/commit/e62630e310ccb0f7ea173137dc30a72406b8bc50))
* use koma length for draw frames ([e37ff01](https://github.com/anydown/mangadown/commit/e37ff0182fc7ed1205c20fbe6a605d5b7e418a51))

<a name="1.1.0"></a>
# 1.1.0 (2017-06-30)


### Bug Fixes

* Add support devicePixelRatio > 1 ([3d939d5](https://github.com/anydown/mangadown/commit/3d939d5))
* discard selection when export image ([344a482](https://github.com/anydown/mangadown/commit/344a482))
* fix image url ([fe056ed](https://github.com/anydown/mangadown/commit/fe056ed))
* remove multiple objects ([c41b7b5](https://github.com/anydown/mangadown/commit/c41b7b5))
* use koma length for draw frames ([e37ff01](https://github.com/anydown/mangadown/commit/e37ff01))


### Features

* add freedrawing mode ([5d3b84f](https://github.com/anydown/mangadown/commit/5d3b84f))
* add local image loader ([c3e2c23](https://github.com/anydown/mangadown/commit/c3e2c23))
* fill white inner bubbles ([42b9ff8](https://github.com/anydown/mangadown/commit/42b9ff8))



<a name="1.0.0"></a>
# 1.0.0 (2017-06-27)


### Bug Fixes

* discard selection when export image ([344a482](https://github.com/anydown/mangadown/commit/344a482))
* fix image url ([fe056ed](https://github.com/anydown/mangadown/commit/fe056ed))
* remove multiple objects ([c41b7b5](https://github.com/anydown/mangadown/commit/c41b7b5))
* use koma length for draw frames ([e37ff01](https://github.com/anydown/mangadown/commit/e37ff01))


### Features

* add freedrawing mode ([5d3b84f](https://github.com/anydown/mangadown/commit/5d3b84f))
* add local image loader ([c3e2c23](https://github.com/anydown/mangadown/commit/c3e2c23))
* fill white inner bubbles ([42b9ff8](https://github.com/anydown/mangadown/commit/42b9ff8))


index.js
var example = `このツールは
Twitterで
よく流れている
タイプの４コマを
作るツールだ…

テキストを編集すると
セリフが変更されるんだね
----
白ハゲスタンプも
押すことが出来る

なるほど
---
よーし

このツールで
主語の大きい
煽り画像を作るぞ！
---
そういうことは
やめような

平和は守られた
`

function isFirefox() {
  return navigator.userAgent.toLowerCase().includes("firefox");
}

function tategaki(ctx, text, x, y, px, font) {
  ctx.fillStyle = ""
  ctx.font = px + `px '${font}'`;
  for (var i = 0; i < text.length; i++) {
    ctx.save();
    ctx.translate(x, y + px * i);
    if (isFirefox()) {
      ctx.translate(10, -px);
    } else {
      ctx.rotate(Math.PI / 2);
      ctx.translate(2, -px - 3);
    }
    ctx.fillText(text[i], 0, px);
    ctx.restore();
  }
}

function drawTextLines(ctx, text, x, y, fontsize, font) {
  var lines = text.split("\n");
  var sx = x + fontsize
  var sy = y
  var maxlen = 0
  for (var l = 0; l < lines.length; l++) {
    var line = lines[l]
    maxlen = maxlen < line.length ? line.length : maxlen
  }
  var cx = sx - fontsize * lines.length / 2
  var cy = sy + fontsize * maxlen / 2

  ctx.save();
  ctx.beginPath();
  ctx.fillStyle = "white"
  ctx.ellipse(cx + fontsize / 8, cy, fontsize * lines.length / 2 + fontsize * 2, fontsize * maxlen / 2 + fontsize * 2, 0, 0, 2 * Math.PI);
  ctx.fill();
  ctx.stroke();
  ctx.restore();

  for (var l = 0; l < lines.length; l++) {
    var line = lines[l]
    tategaki(ctx, line, x - fontsize * l, y, fontsize, font);
  }
}

var fab
var w = 420 + 20 * 2
var h = (280 + 20) * 4 + 20

var loadedFonts = {}

new Vue({
  el: "#app",
  data: {
    msg: "Hello",
    rawinput: example,
    ctx: undefined,
    isDrawingMode: false,
    font: ""
  },
  computed: {
    input: function () {
      var lines = this.rawinput.split("\n")
      var komas = []
      var koma = []
      var msg = []
      for (var i = 0; i < lines.length; i++) {
        var line = lines[i]
        //空行
        if (line.trim() === "" && msg.length > 0) {
          koma.push(msg.join("\n"))
          msg = []
          continue
        }

        //改ページ
        if (line.match(/-+/)) {
          if (msg.length > 0) {
            koma.push(msg.join("\n"))
            msg = []
          }
          komas.push(koma)
          koma = []
          continue
        }

        msg.push(line)
      }
      if (koma.length > 0) {
        komas.push(koma)
      }
      return komas
    }
  },
  methods: {
    toggleDrawingMode: function () {
      this.isDrawingMode = !this.isDrawingMode
      fab.isDrawingMode = this.isDrawingMode;
      fab.freeDrawingBrush.width = 5;
    },
    addImage: function (img) {
      fabric.Image.fromURL(`./${img}.png`, function (img) {
        img.set('left', 100).set('top', 100)
        fab.add(img);
      });
    },
    removeObj: function () {
        if (fab.getActiveGroup()) {
          fab.getActiveGroup().forEachObject(function(o){ fab.remove(o) });
          fab.discardActiveGroup().renderAll();
        }
        if (fab.getActiveObject()) {
          fab.getActiveObject().remove();
        }
    },
    exportImage: function () {
      fab.discardActiveObject().renderAll();
      fab.discardActiveGroup().renderAll();
      var eimage = document.querySelector("#export")
      var canvas = document.querySelector("#output")
      var fabel = document.querySelector("#overlay")
      var destCtx = eimage.getContext('2d');
      destCtx.fillStyle = "white"
      destCtx.fillRect(0, 0, w, h)
      destCtx.drawImage(canvas, 0, 0);
      destCtx.drawImage(fabel, 0, 0, w, h);
      window.open(eimage.toDataURL('image/png'));
    },
    redraw: function () {
      var text = this.input
      var ctx = this.ctx
      ctx.clearRect(0, 0, w, h)

      for (var i = 0; i < text.length; i++) {
        var ox = 20
        var oy = 20 + (280 + 20) * i
        ctx.strokeStyle = "black"
        ctx.lineWidth = 4
        ctx.strokeRect(ox, oy, 420, 280)

        var fontsize = 18;
        var x
        var y = oy + 30
        if (text[i]) {
          if (text[i].length > 0) {
            x = ox + 400 - fontsize
            drawTextLines(ctx, text[i][0], x, y, fontsize, this.font)
          }
          if (text[i].length > 1) {
            var linelength = text[i][1].split("\n").length;
            x = ox + 20 + fontsize * (linelength - 1)
            drawTextLines(ctx, text[i][1], x, y, fontsize, this.font)
          }
        }
      }
    }
  },
  watch: {
    "input": function () {
      this.redraw()
    },
    "font": function(item){
      var self = this
      var fontName = item

      if(!loadedFonts[item]){
        var fonts = {
          "yasashisa": "url(07YasashisaAntique.otf)",
          "noto-tc": "url(NotoSerifTC-Regular.otf)"
        }

        var font = new FontFace(item, fonts[item], {});
        font.load().then((loadedFace) => {
          setTimeout(() => {
            document.fonts.add(loadedFace);
            console.log("font loaded")
            self.redraw();
          }, 1000);
        });
      }

      setTimeout(() => {
        self.redraw();
      }, 1000);
      //var font = new FontFace("newfont", "url(07YasashisaAntique.otf)", {});
    }
  },
  mounted: function () {
    var eimage = document.querySelector("#export")
    eimage.width = `${w}`
    eimage.height = `${h}`
    eimage.style.width = `${w}px`
    eimage.style.height = `${h}px`

    var canvas = document.querySelector("#output")

    canvas.width = `${w}`
    canvas.height = `${h}`
    canvas.style.width = `${w}px`
    canvas.style.height = `${h}px`
    this.ctx = canvas.getContext("2d")

    var fabel = document.querySelector("#overlay")
    fabel.width = `${w}`
    fabel.height = `${h}`
    fabel.style.width = `${w}px`
    fabel.style.height = `${h}px`
    fab = new fabric.Canvas('overlay');

    window.addEventListener('keyup', (ev) => {
      if (ev.keyCode === 46) {
        if (fab.getActiveGroup()) {
          fab.getActiveGroup().forEachObject(function(o){ fab.remove(o) });
          fab.discardActiveGroup().renderAll();
        }
        if (fab.getActiveObject()) {
          fab.getActiveObject().remove();
        }
      }
    })

    document.getElementById('imgLoader').onchange = function handleImage(e) {
      var reader = new FileReader();
      reader.onload = function (event) {
        var imgObj = new Image();
        imgObj.src = event.target.result;
        imgObj.onload = function () {
          var image = new fabric.Image(imgObj);
          image.set({
            height: 200,
            width: 200
          });
          fab.add(image);
          fab.renderAll();
        }
      }
      reader.readAsDataURL(e.target.files[0]);
    }

    this.font = "yasashisa"
  }
})



src/components/draw.js

function isFirefox() {
    return navigator.userAgent.toLowerCase().includes("firefox");
}

export function tategaki(ctx, text, x, y, px, font) {
    ctx.fillStyle = "";
    ctx.font = px + `px '${font}'`;
    for (let i = 0; i < text.length; i++) {
        // 縦書きのために一文字ずつtranslateをかけている
        ctx.save();
        ctx.translate(x, y + px * i);
        if (isFirefox()) {
            ctx.translate(10, -px);
        } else {
            ctx.rotate(Math.PI / 2);
            ctx.translate(2, -px - 3);
        }
        ctx.fillText(text[i], 0, px);
        ctx.restore();
        // かならずrestoreでtranslateをクリアする
    }
}

export function drawBubble(ctx, text, x, y, fontsize, font) {
    const lines = text.split("\n");
    const sx = x + fontsize;
    const sy = y;
    let maxlen = 0;
    for (let l = 0; l < lines.length; l++) {
        const line = lines[l];
        maxlen = maxlen < line.length ? line.length : maxlen;
    }
    const cx = sx - (fontsize * lines.length) / 2;
    const cy = sy + (fontsize * maxlen) / 2;

    ctx.ellipse(
        cx + fontsize / 8,
        cy,
        (fontsize * lines.length) / 2 + fontsize * 2,
        (fontsize * maxlen) / 2 + fontsize * 2,
        0,
        0,
        2 * Math.PI
    );

}

export function drawText(ctx, text, x, y, fontsize, font) {
    const lines = text.split("\n");
    for (let l = 0; l < lines.length; l++) {
        const line = lines[l];
        tategaki(ctx, line, x - fontsize * l, y, fontsize, font);
    }
}

export function drawKomas(ctx, text, font) {
    for (let i = 0; i < text.length; i++) {
        const ox = 20;
        const oy = 20 + (280 + 20) * i;
        ctx.strokeStyle = "black";
        ctx.lineWidth = 4;
        ctx.strokeRect(ox, oy, 420, 280);

        const fontsize = 18;
        let x;
        const y = oy + 30;
        if (text[i]) {
            // 吹き出しの描画
            ctx.save();
            ctx.fillStyle = "white";

            //コマからはみ出さないようにマスク
            ctx.rect(ox + 2, oy + 2, 420 - 4, 280 - 4);
            ctx.clip();
            ctx.beginPath();
            if (text[i].length > 0) {
                x = ox + 400 - fontsize;
                drawBubble(ctx, text[i][0], x, y, fontsize, font);
            }
            ctx.fill();
            ctx.stroke();

            //コマからはみ出さないようにマスク
            ctx.rect(ox + 2, oy + 2, 420 - 4, 280 - 4);
            ctx.clip();
            ctx.beginPath();
            if (text[i].length > 1) {
                const linelength = text[i][1].split("\n").length;
                x = ox + 20 + fontsize * (linelength - 1);
                drawBubble(ctx, text[i][1], x, y, fontsize, font);
            }
            ctx.fill();
            ctx.stroke();
            ctx.restore();

            // セリフの描画
            if (text[i].length > 0) {
                x = ox + 400 - fontsize;
                drawText(ctx, text[i][0], x, y, fontsize, font);
            }
            if (text[i].length > 1) {
                const linelength = text[i][1].split("\n").length;
                x = ox + 20 + fontsize * (linelength - 1);
                drawText(ctx, text[i][1], x, y, fontsize, font);
            }
        }
    }

}

bundle.js
var example = "このツールは\nTwitterで\nよく流れている\nタイプの４コマを\n作るツールだ…\n\nテキストを編集すると\nセリフが変更されるんだね\n----\n白ハゲスタンプも\n押すことが出来る\n\nなるほど\n---\nよーし\n\nこのツールで\n主語の大きい\n煽り画像を作るぞ！\n---\nそういうことは\nやめような\n\n平和は守られた\n"

function isFirefox() {
  return navigator.userAgent.toLowerCase().includes("firefox");
}

function tategaki(ctx, text, x, y, px, font) {
  ctx.fillStyle = ""
  ctx.font = px + "px '" + font + "'";
  for (var i = 0; i < text.length; i++) {
    ctx.save();
    ctx.translate(x, y + px * i);
    if (isFirefox()) {
      ctx.translate(10, -px);
    } else {
      ctx.rotate(Math.PI / 2);
      ctx.translate(2, -px - 3);
    }
    ctx.fillText(text[i], 0, px);
    ctx.restore();
  }
}

function drawTextLines(ctx, text, x, y, fontsize, font) {
  var lines = text.split("\n");
  var sx = x + fontsize
  var sy = y
  var maxlen = 0
  for (var l = 0; l < lines.length; l++) {
    var line = lines[l]
    maxlen = maxlen < line.length ? line.length : maxlen
  }
  var cx = sx - fontsize * lines.length / 2
  var cy = sy + fontsize * maxlen / 2

  ctx.save();
  ctx.beginPath();
  ctx.fillStyle = "white"
  ctx.ellipse(cx + fontsize / 8, cy, fontsize * lines.length / 2 + fontsize * 2, fontsize * maxlen / 2 + fontsize * 2, 0, 0, 2 * Math.PI);
  ctx.fill();
  ctx.stroke();
  ctx.restore();

  for (var l = 0; l < lines.length; l++) {
    var line = lines[l]
    tategaki(ctx, line, x - fontsize * l, y, fontsize, font);
  }
}

var fab
var w = 420 + 20 * 2
var h = (280 + 20) * 4 + 20

var loadedFonts = {}

new Vue({
  el: "#app",
  data: {
    msg: "Hello",
    rawinput: example,
    ctx: undefined,
    isDrawingMode: false,
    font: ""
  },
  computed: {
    input: function () {
      var lines = this.rawinput.split("\n")
      var komas = []
      var koma = []
      var msg = []
      for (var i = 0; i < lines.length; i++) {
        var line = lines[i]
        //空行
        if (line.trim() === "" && msg.length > 0) {
          koma.push(msg.join("\n"))
          msg = []
          continue
        }

        //改ページ
        if (line.match(/-+/)) {
          if (msg.length > 0) {
            koma.push(msg.join("\n"))
            msg = []
          }
          komas.push(koma)
          koma = []
          continue
        }

        msg.push(line)
      }
      if (koma.length > 0) {
        komas.push(koma)
      }
      return komas
    }
  },
  methods: {
    toggleDrawingMode: function () {
      this.isDrawingMode = !this.isDrawingMode
      fab.isDrawingMode = this.isDrawingMode;
      fab.freeDrawingBrush.width = 5;
    },
    addImage: function (img) {
      fabric.Image.fromURL(("./" + img + ".png"), function (img) {
        img.set('left', 100).set('top', 100)
        fab.add(img);
      });
    },
    removeObj: function () {
        if (fab.getActiveGroup()) {
          fab.getActiveGroup().forEachObject(function(o){ fab.remove(o) });
          fab.discardActiveGroup().renderAll();
        }
        if (fab.getActiveObject()) {
          fab.getActiveObject().remove();
        }
    },
    exportImage: function () {
      fab.discardActiveObject().renderAll();
      fab.discardActiveGroup().renderAll();
      var eimage = document.querySelector("#export")
      var canvas = document.querySelector("#output")
      var fabel = document.querySelector("#overlay")
      var destCtx = eimage.getContext('2d');
      destCtx.fillStyle = "white"
      destCtx.fillRect(0, 0, w, h)
      destCtx.drawImage(canvas, 0, 0);
      destCtx.drawImage(fabel, 0, 0, w, h);
      window.open(eimage.toDataURL('image/png'));
    },
    redraw: function () {
      var this$1 = this;

      var text = this.input
      var ctx = this.ctx
      ctx.clearRect(0, 0, w, h)

      for (var i = 0; i < text.length; i++) {
        var ox = 20
        var oy = 20 + (280 + 20) * i
        ctx.strokeStyle = "black"
        ctx.lineWidth = 4
        ctx.strokeRect(ox, oy, 420, 280)

        var fontsize = 18;
        var x
        var y = oy + 30
        if (text[i]) {
          if (text[i].length > 0) {
            x = ox + 400 - fontsize
            drawTextLines(ctx, text[i][0], x, y, fontsize, this$1.font)
          }
          if (text[i].length > 1) {
            var linelength = text[i][1].split("\n").length;
            x = ox + 20 + fontsize * (linelength - 1)
            drawTextLines(ctx, text[i][1], x, y, fontsize, this$1.font)
          }
        }
      }
    }
  },
  watch: {
    "input": function () {
      this.redraw()
    },
    "font": function(item){
      var self = this
      var fontName = item

      if(!loadedFonts[item]){
        var fonts = {
          "yasashisa": "url(07YasashisaAntique.otf)",
          "noto-tc": "url(NotoSerifTC-Regular.otf)"
        }

        var font = new FontFace(item, fonts[item], {});
        font.load().then(function (loadedFace) {
          setTimeout(function () {
            document.fonts.add(loadedFace);
            console.log("font loaded")
            self.redraw();
          }, 1000);
        });
      }

      setTimeout(function () {
        self.redraw();
      }, 1000);
      //var font = new FontFace("newfont", "url(07YasashisaAntique.otf)", {});
    }
  },
  mounted: function () {
    var eimage = document.querySelector("#export")
    eimage.width = "" + w
    eimage.height = "" + h
    eimage.style.width = w + "px"
    eimage.style.height = h + "px"

    var canvas = document.querySelector("#output")

    canvas.width = "" + w
    canvas.height = "" + h
    canvas.style.width = w + "px"
    canvas.style.height = h + "px"
    this.ctx = canvas.getContext("2d")

    var fabel = document.querySelector("#overlay")
    fabel.width = "" + w
    fabel.height = "" + h
    fabel.style.width = w + "px"
    fabel.style.height = h + "px"
    fab = new fabric.Canvas('overlay');

    window.addEventListener('keyup', function (ev) {
      if (ev.keyCode === 46) {
        if (fab.getActiveGroup()) {
          fab.getActiveGroup().forEachObject(function(o){ fab.remove(o) });
          fab.discardActiveGroup().renderAll();
        }
        if (fab.getActiveObject()) {
          fab.getActiveObject().remove();
        }
      }
    })

    document.getElementById('imgLoader').onchange = function handleImage(e) {
      var reader = new FileReader();
      reader.onload = function (event) {
        var imgObj = new Image();
        imgObj.src = event.target.result;
        imgObj.onload = function () {
          var image = new fabric.Image(imgObj);
          image.set({
            height: 200,
            width: 200
          });
          fab.add(image);
          fab.renderAll();
        }
      }
      reader.readAsDataURL(e.target.files[0]);
    }

    this.font = "yasashisa"
  }
})


